File: packages/skillet-eval-python/src/skill_eval/runner/transcript.py
from __future__ import annotations

from dataclasses import dataclass

from skill_eval.providers.types import ToolCall


@dataclass
class ToolResult:
    id: str
    name: str
    result: object


@dataclass
class TranscriptStep:
    step: int
    turn: int | None = None
    user_message: str | None = None
    turn_skipped: str | None = None
    response: str = ""
    tool_calls: list[ToolCall] | None = None
    tool_results: list[ToolResult] = None
    usage: dict[str, int] = None
    latency_ms: float = 0

    def __post_init__(self) -> None:
        if self.tool_results is None:
            self.tool_results = []
        if self.usage is None:
            self.usage = {"inputTokens": 0, "outputTokens": 0}


@dataclass
class AgentRun:
    transcript: list[TranscriptStep]
    final_output: str
    total_input_tokens: int
    total_output_tokens: int
    total_latency_ms: float
    total_tool_calls: int
    errors: int
    steps: int


def build_agent_run(transcript: list[TranscriptStep]) -> AgentRun:
    total_input_tokens = 0
    total_output_tokens = 0
    total_latency_ms = 0.0
    total_tool_calls = 0
    errors = 0

    for step in transcript:
        total_input_tokens += step.usage.get("inputTokens", 0)
        total_output_tokens += step.usage.get("outputTokens", 0)
        total_latency_ms += step.latency_ms
        total_tool_calls += len(step.tool_calls) if step.tool_calls else 0
        for tr in step.tool_results:
            r = tr.result
            if isinstance(r, dict) and r.get("error"):
                errors += 1

    last_step = transcript[-1] if transcript else None
    return AgentRun(
        transcript=transcript,
        final_output=last_step.response if last_step else "",
        total_input_tokens=total_input_tokens,
        total_output_tokens=total_output_tokens,
        total_latency_ms=total_latency_ms,
        total_tool_calls=total_tool_calls,
        errors=errors,
        steps=len(transcript),
    )

File: packages/skillet/src/eval/runner/transcript.ts
import type { ToolCall } from "../providers/types.js";

export interface ToolResult {
	id: string;
	name: string;
	result: unknown;
}

export interface TranscriptStep {
	step: number;
	turn?: number;
	userMessage?: string;
	/** Set when a scripted turn was skipped because it didn't match the agent's response. */
	turnSkipped?: string;
	response: string;
	toolCalls?: ToolCall[];
	toolResults: ToolResult[];
	usage: { inputTokens: number; outputTokens: number };
	latencyMs: number;
}

export interface AgentRun {
	transcript: TranscriptStep[];
	finalOutput: string;
	totalInputTokens: number;
	totalOutputTokens: number;
	totalLatencyMs: number;
	totalToolCalls: number;
	errors: number;
	steps: number;
}

export function buildAgentRun(transcript: TranscriptStep[]): AgentRun {
	let totalInputTokens = 0;
	let totalOutputTokens = 0;
	let totalLatencyMs = 0;
	let totalToolCalls = 0;
	let errors = 0;

	for (const step of transcript) {
		totalInputTokens += step.usage.inputTokens;
		totalOutputTokens += step.usage.outputTokens;
		totalLatencyMs += step.latencyMs;
		totalToolCalls += step.toolCalls?.length ?? 0;
		for (const tr of step.toolResults) {
			const r = tr.result as Record<string, unknown>;
			if (r?.error) errors++;
		}
	}

	const lastStep = transcript[transcript.length - 1];
	return {
		transcript,
		finalOutput: lastStep?.response ?? "",
		totalInputTokens,
		totalOutputTokens,
		totalLatencyMs,
		totalToolCalls,
		errors,
		steps: transcript.length,
	};
}
